fdbclient/BlobCipher.cpp

```diff
@@ -744,20 +798,23 @@ void DecryptBlobCipherAes256Ctr::verifyHeaderMultiAuthToken(const uint8_t* ciphe
 	if (!headerAuthTokenValidationDone) {
 		verifyHeaderAuthToken(header, arena);
 	}
-	uint8_t computedCipherTextAuthToken[AUTH_TOKEN_SIZE];
+	uint8_t computedCipherTextAuthToken[AUTH_TOKEN_MAX_SIZE];
+	// TOOD: Use HMAC_SHA encyrption authentication scheme as AES_CMAC needs minimum 16 bytes cipher key
 	computeAuthToken(ciphertext,
 	                 ciphertextLen,
 	                 reinterpret_cast<const uint8_t*>(&header.cipherTextDetails.salt),
 	                 sizeof(EncryptCipherRandomSalt),
 	                 &computedCipherTextAuthToken[0],
-	                 AUTH_TOKEN_SIZE);
-	if (memcmp(&header.multiAuthTokens.cipherTextAuthToken[0], &computedCipherTextAuthToken[0], AUTH_TOKEN_SIZE) != 0) {
+	                 EncryptAuthTokenAlgo::ENCRYPT_HEADER_AUTH_TOKEN_ALGO_HMAC_SHA,
```

Comment:
Good point. Thanks for pointing it out.

Reply:
Btw, I just realized, for MULTI auth token mode, for the usecase where we use `randomSalt` as `authentication cipherKey`; I'm using HMAC_SHA as auth-digest scheme as CMAC needs 16 bytes of the cipherKey. Also, the code has TODO to cover this usecase:

`// TOOD: Use HMAC_SHA encyrption authentication scheme as AES_CMAC needs minimum 16 bytes cipher key`

Reply:
I didn't realize random salt is used as auth key. Maybe we could do CMAC-AES128 instead of CMAC-AES256. But the bigger question is whether the use of random salt for auth key is reviewed by security expert. In general with a plaintext auth key (which is the salt), an attacker can replace the whole content plus the auth key and have it pass the auth verification. But here the attack is harder, since it is hard to construct a salt to pass the header authentication, and hard to construct a salt+content pair that decrypts to meaningful data. Even so, since the header auth token is readily available, is there reason why we don't use it?

Reply:
> But the bigger question is whether the use of random salt for auth key is reviewed by security expert. 

Yes it is; the reason it is safe is, here we are working on an encrypted `payload` compared to `plaintext header` (where a KMS managed cipher-key needs to be used). So the goal is to have protection against bit-rot and bit-flip as tampering could be detected while we try to decrypt the encrypted buffer. 

> CMAC-AES128 
As mentioned in the TODO, AES-128 needs 16 byte cipherKey, AES_256 needs 32 bytes cipher-key length

I did think about ways to circumvent the situation, valid options are:
1. Covert 8 byte salt to 32/16 byte key using XOR or other concatenation techniques 
2. Replace authToken with xxhash64 bit token 

The main reason I didn't do it (and put a TODO) is the following: MULTI mode motivation was to cover large file encryption usecase such as: BlobGranule and FDB backup file, however, we choose to construct these files as small size `chunks` and encrypt (and/or compress) those chunks individuals, hence, we ended up using SINGLE auth-token mode. I even thought about removing the MULTI mode, but, then left it there for addressing future usecases if any. My take is, lets leave the TODO for now and avoid potential over-engineering; if need arises we can fix this gap in multiple ways.

Reply:
Thinking more it is fair to remove MULTI token support for now, further, it would save 16 bytes from EncyrptionHeader, which in grand scheme of things may seems like a decent saving. 

However, I would like to do that cleanup as a separate PR unless someone has strong opinions about it. 

Thanks

Reply:
sounds good.